harness/determined/pytorch/_trainer.py

```diff
@@ -176,11 +175,37 @@ def fit(
 
             smaller_is_better = bool(self._info.trial._config["searcher"]["smaller_is_better"])
             searcher_metric_name = self._info.trial._config["searcher"]["metric"]
+
             steps_completed = int(self._info.trial._steps_completed)
             global_batch_size = self._info.trial.hparams.get("global_batch_size", None)
             if global_batch_size:
                 global_batch_size = int(global_batch_size)
 
+            # Backwards compatibility: try to parse legacy `searcher.max_length` if `max_length`
+            # isn't passed in.
+            if max_length is None:
+                max_length_val = core._parse_searcher_max_length(self._info.trial._config)
+                if max_length_val:
+                    logger.warning(
+                        "Configuring `max_length` from the `searcher.max_length` experiment config,"
+                        " which was deprecated in XXYYZZ and will be removed in a future release. "
+                        "Please set `fit(max_length=X)` with your desired training length directly."
+                    )
+                    max_length_unit = core._parse_searcher_units(self._info.trial._config)
+                    max_length = pytorch.TrainUnit._from_searcher_unit(
+                        max_length_val, max_length_unit, global_batch_size
+                    )
+
+            # If we couldn't parse the legacy `searcher.max_length`, raise an error.
+            if not max_length:
+                raise ValueError(
+                    "`fit(max_length=X)` must be set with your desired training length."
+                )
+            if not isinstance(max_length.value, int):
+                raise TypeError("max_length must be configured in TrainUnit(int) types.")
```

Comment:
Can you say `max_length must either be a Batch() or Epoch() type` instead?

Not sure that the `TrainUnit` abstract type will make immediate sense to readers.